File: 127-word-ladder/word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> st(wordList.begin(), wordList.end());
        if (st.find(endWord) == st.end()) return 0;
        queue<pair<string, int>> q;
        q.push({beginWord, 1});
        st.erase(beginWord); 

        while (!q.empty()) {
            string word = q.front().first;
            int steps = q.front().second;
            q.pop();

            for (int i = 0; i < word.size(); i++) {
                char originalChar = word[i]; 

                for (char ch = 'a'; ch <= 'z'; ch++) {
                    word[i] = ch;

                    if (word == endWord) return steps + 1;

                    if (st.find(word) != st.end()) {
                        st.erase(word);
                        q.push({word, steps + 1});
                    }
                }

                word[i] = originalChar; 
            }
        }

        return 0; 
    }
};

File: 127-word-ladder/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        visited=set(wordList)
        q=deque()
        q.append([beginWord,1])
        while q:
            word,steps=q.popleft()
            for i in range(len(word)):
                for ch in string.ascii_lowercase:
                    newword=word[:i]+ch+word[i+1:]

                    if newword==endWord:
                        return steps+1
                    
                    if newword in visited:
                        visited.remove(newword)
                        q.append([newword,steps+1])

        return 0